File: src/lib/pjecalc/core/dominio/calculo/ferias/ferias.ts
/**
 * Porte 1:1 de Ferias.java (472 linhas).
 *
 * Entidade de férias. Cada instância representa UM período aquisitivo + concessivo
 * + até 3 períodos de gozo + abono + dobras.
 *
 * Campos principais:
 *   - periodoAquisitivo (12 meses do contrato gerando direito)
 *   - periodoConcessivo (12 meses após aquisitivo para usufruir)
 *   - prazo (dias a gozar — padrão 30)
 *   - situacao (GOZADAS, INDENIZADAS, GOZADAS_PARCIALMENTE, PERDIDAS)
 *   - até 3 períodos de gozo (inicial, final, dobra) — permite gozo fracionado
 *   - abono pecuniário (1/3 dos dias vendidos)
 *
 * Ref: pjecalc-fonte/.../dominio/calculo/ferias/Ferias.java
 */
import { Periodo } from '../../../base/comum/periodo';
import { HelperDate } from '../../../base/comum/helper-date';
import { SituacaoDaFeriasEnum } from '../../../constantes/enums';
import type { Calculo } from '../calculo';
import type { ParametroDoTermo } from '../../termo/parametro-do-termo';

export { SituacaoDaFeriasEnum };

export class Ferias {
  private static readonly DIAS_ABONO_PADRAO = 10;

  private id: number | null = null;
  private versao: number = 0;
  private calculo: Calculo | null = null;
  private relativa: string = '';
  private dataInicialDoPeriodoAquisitivo: Date | null = null;
  private dataFinalDoPeriodoAquisitivo: Date | null = null;
  private dataInicialDoPeriodoConcessivo: Date | null = null;
  private dataFinalDoPeriodoConcessivo: Date | null = null;
  private prazo: number = 30;
  private situacao: SituacaoDaFeriasEnum = SituacaoDaFeriasEnum.GOZADAS;
  private dobraGeral: boolean = false;
  private abono: boolean = false;
  private quantidadeDiasAbono: number = Ferias.DIAS_ABONO_PADRAO;
  private dataInicialDoPeriodoDeGozo1: Date | null = null;
  private dataFinalDoPeriodoDeGozo1: Date | null = null;
  private dobraDoPeriodoDeGozo1: boolean = false;
  private dataInicialDoPeriodoDeGozo2: Date | null = null;
  private dataFinalDoPeriodoDeGozo2: Date | null = null;
  private dobraDoPeriodoDeGozo2: boolean = false;
  private dataInicialDoPeriodoDeGozo3: Date | null = null;
  private dataFinalDoPeriodoDeGozo3: Date | null = null;
  private dobraDoPeriodoDeGozo3: boolean = false;

  // Transient caches
  private periodoAquisitivo: Periodo | null = null;
  private periodoConcessivo: Periodo | null = null;

  getId(): number | null { return this.id; }
  setId(id: number | null): void { this.id = id; }
  getVersao(): number { return this.versao; }
  setVersao(v: number): void { this.versao = v; }

  setPeriodoAquisitivo(p: Periodo): void {
    this.periodoAquisitivo = p;
    this.setDataInicialDoPeriodoAquisitivo(p.getInicial());
    this.setDataFinalDoPeriodoAquisitivo(p.getFinal());
  }

  getPeriodoAquisitivo(): Periodo {
    if (!this.periodoAquisitivo) {
      this.periodoAquisitivo = new Periodo(
        this.dataInicialDoPeriodoAquisitivo, this.dataFinalDoPeriodoAquisitivo,
      );
    }
    return this.periodoAquisitivo;
  }

  getPeriodoDeGozo1(): Periodo | null {
    if (!this.dataInicialDoPeriodoDeGozo1 || !this.dataFinalDoPeriodoDeGozo1) return null;
    return new Periodo(this.dataInicialDoPeriodoDeGozo1, this.dataFinalDoPeriodoDeGozo1);
  }

  getPeriodoDeGozo2(): Periodo | null {
    if (!this.dataInicialDoPeriodoDeGozo2 || !this.dataFinalDoPeriodoDeGozo2) return null;
    return new Periodo(this.dataInicialDoPeriodoDeGozo2, this.dataFinalDoPeriodoDeGozo2);
  }

  getPeriodoDeGozo3(): Periodo | null {
    if (!this.dataInicialDoPeriodoDeGozo3 || !this.dataFinalDoPeriodoDeGozo3) return null;
    return new Periodo(this.dataInicialDoPeriodoDeGozo3, this.dataFinalDoPeriodoDeGozo3);
  }

  setPeriodoConcessivo(p: Periodo): void {
    this.periodoConcessivo = p;
    this.setDataInicialDoPeriodoConcessivo(p.getInicial());
    this.setDataFinalDoPeriodoConcessivo(p.getFinal());
  }

  getPeriodoConcessivo(): Periodo {
    if (!this.periodoConcessivo) {
      this.periodoConcessivo = new Periodo(
        this.dataInicialDoPeriodoConcessivo, this.dataFinalDoPeriodoConcessivo,
      );
    }
    return this.periodoConcessivo;
  }

  setPeriodoDeGozo1(p: Periodo): void {
    this.setDataInicialDoPeriodoDeGozo1(p.getInicial());
    this.setDataFinalDoPeriodoDeGozo1(p.getFinal());
  }
  setPeriodoDeGozo2(p: Periodo): void {
    this.setDataInicialDoPeriodoDeGozo2(p.getInicial());
    this.setDataFinalDoPeriodoDeGozo2(p.getFinal());
  }
  setPeriodoDeGozo3(p: Periodo): void {
    this.setDataInicialDoPeriodoDeGozo3(p.getInicial());
    this.setDataFinalDoPeriodoDeGozo3(p.getFinal());
  }

  getCalculo(): Calculo | null { return this.calculo; }
  setCalculo(c: Calculo | null): void { this.calculo = c; }

  getRelativa(): string { return this.relativa; }
  setRelativa(v: string): void { this.relativa = v; }

  getDataInicialDoPeriodoAquisitivo(): Date | null { return this.dataInicialDoPeriodoAquisitivo; }
  setDataInicialDoPeriodoAquisitivo(d: Date | null): void { this.dataInicialDoPeriodoAquisitivo = d; }
  getDataFinalDoPeriodoAquisitivo(): Date | null { return this.dataFinalDoPeriodoAquisitivo; }
  setDataFinalDoPeriodoAquisitivo(d: Date | null): void { this.dataFinalDoPeriodoAquisitivo = d; }

  getDataInicialDoPeriodoConcessivo(): Date | null { return this.dataInicialDoPeriodoConcessivo; }
  setDataInicialDoPeriodoConcessivo(d: Date | null): void { this.dataInicialDoPeriodoConcessivo = d; }
  getDataFinalDoPeriodoConcessivo(): Date | null { return this.dataFinalDoPeriodoConcessivo; }
  setDataFinalDoPeriodoConcessivo(d: Date | null): void { this.dataFinalDoPeriodoConcessivo = d; }

  getAbono(): boolean { return this.abono; }
  setAbono(v: boolean): void { this.abono = v; }
  getQuantidadeDiasAbono(): number { return this.quantidadeDiasAbono; }
  setQuantidadeDiasAbono(v: number): void { this.quantidadeDiasAbono = v; }

  getSituacao(): SituacaoDaFeriasEnum { return this.situacao; }
  setSituacao(s: SituacaoDaFeriasEnum): void { this.situacao = s; }

  getDataInicialDoPeriodoDeGozo1(): Date | null { return this.dataInicialDoPeriodoDeGozo1; }
  setDataInicialDoPeriodoDeGozo1(d: Date | null): void { this.dataInicialDoPeriodoDeGozo1 = d; }
  getDataFinalDoPeriodoDeGozo1(): Date | null { return this.dataFinalDoPeriodoDeGozo1; }
  setDataFinalDoPeriodoDeGozo1(d: Date | null): void { this.dataFinalDoPeriodoDeGozo1 = d; }

  getDataInicialDoPeriodoDeGozo2(): Date | null { return this.dataInicialDoPeriodoDeGozo2; }
  setDataInicialDoPeriodoDeGozo2(d: Date | null): void { this.dataInicialDoPeriodoDeGozo2 = d; }
  getDataFinalDoPeriodoDeGozo2(): Date | null { return this.dataFinalDoPeriodoDeGozo2; }
  setDataFinalDoPeriodoDeGozo2(d: Date | null): void { this.dataFinalDoPeriodoDeGozo2 = d; }

  getDataInicialDoPeriodoDeGozo3(): Date | null { return this.dataInicialDoPeriodoDeGozo3; }
  setDataInicialDoPeriodoDeGozo3(d: Date | null): void { this.dataInicialDoPeriodoDeGozo3 = d; }
  getDataFinalDoPeriodoDeGozo3(): Date | null { return this.dataFinalDoPeriodoDeGozo3; }
  setDataFinalDoPeriodoDeGozo3(d: Date | null): void { this.dataFinalDoPeriodoDeGozo3 = d; }

  getPrazo(): number { return this.prazo; }
  setPrazo(v: number): void { this.prazo = v; }

  getDobraGeral(): boolean { return this.dobraGeral; }
  setDobraGeral(v: boolean): void { this.dobraGeral = v; }

  getDobraDoPeriodoDeGozo1(): boolean { return this.dobraDoPeriodoDeGozo1; }
  setDobraDoPeriodoDeGozo1(v: boolean): void { this.dobraDoPeriodoDeGozo1 = v; }
  getDobraDoPeriodoDeGozo2(): boolean { return this.dobraDoPeriodoDeGozo2; }
  setDobraDoPeriodoDeGozo2(v: boolean): void { this.dobraDoPeriodoDeGozo2 = v; }
  getDobraDoPeriodoDeGozo3(): boolean { return this.dobraDoPeriodoDeGozo3; }
  setDobraDoPeriodoDeGozo3(v: boolean): void { this.dobraDoPeriodoDeGozo3 = v; }

  /** limparPeriodosDeGozos (Ferias.java:410-417) */
  limparPeriodosDeGozos(): void {
    this.setDataInicialDoPeriodoDeGozo1(null);
    this.setDataFinalDoPeriodoDeGozo1(null);
    this.setDataInicialDoPeriodoDeGozo2(null);
    this.setDataFinalDoPeriodoDeGozo2(null);
    this.setDataInicialDoPeriodoDeGozo3(null);
    this.setDataFinalDoPeriodoDeGozo3(null);
  }

  /**
   * sugerirPrazo (Ferias.java:419-425) — sugere prazo baseado em faltas.
   * STUB: requer CalculoDoPrazoDeFerias (não portado ainda, Fase 10).
   */
  sugerirPrazo(): void {
    // Mantém prazo atual; será completado quando CalculoDoPrazoDeFerias estiver portado.
  }

  /**
   * sugerirPrimeiroPeriodosDeGozos (Ferias.java:427-437).
   */
  sugerirPrimeiroPeriodosDeGozos(): void {
    const calc = this.calculo as unknown as {
      getInicioFeriasColetivas?: () => Date | null;
    };
    const inicioColetivas = calc?.getInicioFeriasColetivas?.();

    if (inicioColetivas && this.dataFinalDoPeriodoAquisitivo
        && HelperDate.dateBefore(this.dataFinalDoPeriodoAquisitivo, inicioColetivas)) {
      const dataInicial = inicioColetivas;
      const periodoDeGozo = new Periodo(
        dataInicial,
        HelperDate.getInstance(dataInicial)!.addDay(this.prazo - 1).getDate(),
      );
      this.setPeriodoDeGozo1(periodoDeGozo);
    } else if (this.situacao === SituacaoDaFeriasEnum.GOZADAS
        && this.dataFinalDoPeriodoConcessivo && this.prazo > 0) {
      const dataInicial = HelperDate.getInstance(this.dataFinalDoPeriodoConcessivo)!
        .addDay(-(this.prazo - 1)).getDate();
      const periodoDeGozo = new Periodo(dataInicial, this.dataFinalDoPeriodoConcessivo);
      this.setPeriodoDeGozo1(periodoDeGozo);
    }
  }

  /**
   * Calcula dias a indenizar (dias do prazo - dias já gozados - dias de abono).
   * Usado em ComportamentoDaBaseDoReflexo.verificarValorParaFerias.
   */
  obterDiasAIndenizar(): number {
    let totalDeDias = this.prazo;
    const g1 = this.getPeriodoDeGozo1();
    const g2 = this.getPeriodoDeGozo2();
    const g3 = this.getPeriodoDeGozo3();
    if (g1) totalDeDias -= g1.totalDeDias();
    if (g2) totalDeDias -= g2.totalDeDias();
    if (g3) totalDeDias -= g3.totalDeDias();
    if (this.abono) totalDeDias -= this.quantidadeDiasAbono;
    return Math.max(0, totalDeDias);
  }

  /**
   * encontrarPrazoFeriasProporcionais (Ferias.java:462-470, static).
   * STUB: requer CalculoDoPrazoDeFerias. Fallback: 30 dias.
   */
  static encontrarPrazoFeriasProporcionais(parametro: ParametroDoTermo): number {
    const calc = parametro.getCalculo() as unknown as {
      getPrazoFeriasProporcional?: () => number | null;
    };
    const prazo = calc?.getPrazoFeriasProporcional?.();
    if (prazo !== null && prazo !== undefined) return prazo;
    return 30;
  }
}

File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/ferias/Ferias.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  javax.persistence.Column
 *  javax.persistence.Entity
 *  javax.persistence.FetchType
 *  javax.persistence.GeneratedValue
 *  javax.persistence.GenerationType
 *  javax.persistence.Id
 *  javax.persistence.JoinColumn
 *  javax.persistence.ManyToOne
 *  javax.persistence.SequenceGenerator
 *  javax.persistence.Table
 *  javax.persistence.Temporal
 *  javax.persistence.TemporalType
 *  javax.persistence.Transient
 *  javax.persistence.Version
 *  org.hibernate.annotations.Parameter
 *  org.hibernate.annotations.Type
 *  org.hibernate.validator.NotNull
 *  org.jboss.seam.annotations.Name
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.ferias;

import br.jus.trt8.pjecalc.base.comum.EntidadeBase;
import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Periodo;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.GerenciadorDeValidadores;
import br.jus.trt8.pjecalc.negocio.comum.rotinasdecalculo.CalculoDoPrazoDeFerias;
import br.jus.trt8.pjecalc.negocio.comum.validators.Required;
import br.jus.trt8.pjecalc.negocio.comum.validators.ValidValue;
import br.jus.trt8.pjecalc.negocio.constantes.SituacaoDaFeriasEnum;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.Calculo;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.ferias.RepositorioDeFerias;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.ferias.regras.AbonoDeFeriasValidRule;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.ferias.regras.DiasDeAbonoValidRule;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.ferias.regras.PeriodoDeGozoValidRule;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.ferias.regras.PrazoDeFeriasValidRule;
import br.jus.trt8.pjecalc.negocio.dominio.termo.ParametroDoTermo;
import java.io.Serializable;
import java.util.Date;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.persistence.Transient;
import javax.persistence.Version;
import org.hibernate.annotations.Parameter;
import org.hibernate.annotations.Type;
import org.hibernate.validator.NotNull;
import org.jboss.seam.annotations.Name;

@Entity
@Table(name="TBFERIASCALCULO")
@SequenceGenerator(name="SQFERIASCALCULO", sequenceName="SQFERIASCALCULO", allocationSize=1)
@Name(value="ferias")
public class Ferias
extends EntidadeBase
implements Serializable {
    private static final long serialVersionUID = -4841534726685343431L;
    private static final Integer DIAS_ABONO_PADRAO = 10;
    @Id
    @GeneratedValue(strategy=GenerationType.SEQUENCE, generator="SQFERIASCALCULO")
    @Column(name="IIDFERIASCALCULO")
    private final Long id = null;
    @Version
    @Column(name="IIDVERSAO")
    private Long versao;
    @ManyToOne(fetch=FetchType.LAZY)
    @JoinColumn(name="IIDCALCULO")
    private Calculo calculo;
    @Column(name="SNMPERIODOFERIAS")
    @Required
    private String relativa;
    @Column(name="DDTINICIOPERIODOAQUISITIVO")
    @Temporal(value=TemporalType.DATE)
    @Required
    private Date dataInicialDoPeriodoAquisitivo;
    @Column(name="DDTTERMINOPERIODOAQUISITIVO")
    @Temporal(value=TemporalType.DATE)
    @Required
    private Date dataFinalDoPeriodoAquisitivo;
    @Column(name="DDTINICIOPERIODOCONCESSIVO")
    @Temporal(value=TemporalType.DATE)
    @Required
    private Date dataInicialDoPeriodoConcessivo;
    @Column(name="DDTTERMINOPERIODOCONCESSIVO")
    @Temporal(value=TemporalType.DATE)
    @Required
    private Date dataFinalDoPeriodoConcessivo;
    @Column(name="IQTDIASFERIAS")
    @ValidValue(validRule=PrazoDeFeriasValidRule.class)
    @Required
    private Integer prazo = 30;
    @Column(name="STPSITUACAOFERIAS", columnDefinition="VARCHAR2(2)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.EnumUserType", parameters={@Parameter(name="enum", value="SituacaoDaFeriasEnum")})
    @NotNull
    private SituacaoDaFeriasEnum situacao = SituacaoDaFeriasEnum.GOZADAS;
    @Column(name="SFLDOBRAGERAL", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.BooleanUserType")
    @Required
    private Boolean dobraGeral = false;
    @Column(name="SFLABONOFERIAS", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.BooleanUserType")
    @ValidValue(validRule=AbonoDeFeriasValidRule.class)
    @Required
    private Boolean abono = false;
    @Column(name="IQTDIASABONO")
    @ValidValue(validRule=DiasDeAbonoValidRule.class)
    @Required
    private Integer quantidadeDiasAbono = DIAS_ABONO_PADRAO;
    @Column(name="DDTINICIOPRIMEIROPERIODO")
    @Temporal(value=TemporalType.DATE)
    @ValidValue(validRule=PeriodoDeGozoValidRule.class, flag=0)
    private Date dataInicialDoPeriodoDeGozo1;
    @Column(name="DDTTERMINOPRIMEIROPERIODO")
    @Temporal(value=TemporalType.DATE)
    @ValidValue(validRule=PeriodoDeGozoValidRule.class, flag=1)
    private Date dataFinalDoPeriodoDeGozo1;
    @Column(name="SFLDOBRAPRIMEIROPERIODO", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.BooleanUserType")
    @Required
    private Boolean dobraDoPeriodoDeGozo1 = false;
    @Column(name="DDTINICIOSEGUNDOPERIODO")
    @Temporal(value=TemporalType.DATE)
    @ValidValue(validRule=PeriodoDeGozoValidRule.class, flag=2)
    private Date dataInicialDoPeriodoDeGozo2;
    @Column(name="DDTTERMINOSEGUNDOPERIODO")
    @Temporal(value=TemporalType.DATE)
    @ValidValue(validRule=PeriodoDeGozoValidRule.class, flag=3)
    private Date dataFinalDoPeriodoDeGozo2;
    @Column(name="SFLDOBRASEGUNDOPERIODO", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.BooleanUserType")
    @Required
    private Boolean dobraDoPeriodoDeGozo2 = false;
    @Column(name="DDTINICIOTERCEIROPERIODO")
    @Temporal(value=TemporalType.DATE)
    @ValidValue(validRule=PeriodoDeGozoValidRule.class, flag=4)
    private Date dataInicialDoPeriodoDeGozo3;
    @Column(name="DDTTERMINOTERCEIROPERIODO")
    @Temporal(value=TemporalType.DATE)
    @ValidValue(validRule=PeriodoDeGozoValidRule.class, flag=5)
    private Date dataFinalDoPeriodoDeGozo3;
    @Column(name="SFLDOBRATERCEIROPERIODO", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.BooleanUserType")
    @Required
    private Boolean dobraDoPeriodoDeGozo3 = false;
    @Transient
    private Periodo periodoAquisitivo;
    @Transient
    private Periodo periodoConcessivo;

    public Ferias() {
        super(RepositorioDeFerias.class);
    }

    @Override
    public Object obterChavePrimaria() {
        return this.getId();
    }

    @Override
    public Long getVersao() {
        return this.versao;
    }

    @Override
    public void setVersao(Long versao) {
        this.versao = versao;
    }

    public Long getId() {
        return this.id;
    }

    public void setPeriodoAquisitivo(Periodo periodo) {
        this.periodoAquisitivo = periodo;
        this.setDataInicialDoPeriodoAquisitivo(periodo.getInicial());
        this.setDataFinalDoPeriodoAquisitivo(periodo.getFinal());
    }

    public Periodo getPeriodoAquisitivo() {
        if (Utils.nulo(this.periodoAquisitivo)) {
            this.periodoAquisitivo = new Periodo(this.getDataInicialDoPeriodoAquisitivo(), this.getDataFinalDoPeriodoAquisitivo());
        }
        return this.periodoAquisitivo;
    }

    public Periodo getPeriodoDeGozo1() {
        return new Periodo(this.getDataInicialDoPeriodoDeGozo1(), this.getDataFinalDoPeriodoDeGozo1());
    }

    public Periodo getPeriodoDeGozo2() {
        return new Periodo(this.getDataInicialDoPeriodoDeGozo2(), this.getDataFinalDoPeriodoDeGozo2());
    }

    public Periodo getPeriodoDeGozo3() {
        return new Periodo(this.getDataInicialDoPeriodoDeGozo3(), this.getDataFinalDoPeriodoDeGozo3());
    }

    public void setPeriodoConcessivo(Periodo periodo) {
        this.periodoConcessivo = periodo;
        this.setDataInicialDoPeriodoConcessivo(periodo.getInicial());
        this.setDataFinalDoPeriodoConcessivo(periodo.getFinal());
    }

    public Periodo getPeriodoConcessivo() {
        if (Utils.nulo(this.periodoConcessivo)) {
            this.periodoConcessivo = new Periodo(this.getDataInicialDoPeriodoConcessivo(), this.getDataFinalDoPeriodoConcessivo());
        }
        return this.periodoConcessivo;
    }

    public void setPeriodoDeGozo1(Periodo periodo) {
        this.setDataInicialDoPeriodoDeGozo1(periodo.getInicial());
        this.setDataFinalDoPeriodoDeGozo1(periodo.getFinal());
    }

    public void setPeriodoDeGozo2(Periodo periodo) {
        this.setDataInicialDoPeriodoDeGozo2(periodo.getInicial());
        this.setDataFinalDoPeriodoDeGozo2(periodo.getFinal());
    }

    public void setPeriodoDeGozo3(Periodo periodo) {
        this.setDataInicialDoPeriodoDeGozo3(periodo.getInicial());
        this.setDataFinalDoPeriodoDeGozo3(periodo.getFinal());
    }

    public Calculo getCalculo() {
        return this.calculo;
    }

    public void setCalculo(Calculo calculo) {
        this.calculo = calculo;
    }

    public String getRelativa() {
        return this.relativa;
    }

    public void setRelativa(String relativa) {
        this.relativa = relativa;
    }

    public Date getDataInicialDoPeriodoAquisitivo() {
        return this.dataInicialDoPeriodoAquisitivo;
    }

    public void setDataInicialDoPeriodoAquisitivo(Date dataInicialDoPeriodoAquisitivo) {
        this.dataInicialDoPeriodoAquisitivo = dataInicialDoPeriodoAquisitivo;
    }

    public Date getDataFinalDoPeriodoAquisitivo() {
        return this.dataFinalDoPeriodoAquisitivo;
    }

    public void setDataFinalDoPeriodoAquisitivo(Date dataFinalDoPeriodoAquisitivo) {
        this.dataFinalDoPeriodoAquisitivo = dataFinalDoPeriodoAquisitivo;
    }

    public Date getDataInicialDoPeriodoConcessivo() {
        return this.dataInicialDoPeriodoConcessivo;
    }

    public void setDataInicialDoPeriodoConcessivo(Date dataInicialDoPeriodoConcessivo) {
        this.dataInicialDoPeriodoConcessivo = dataInicialDoPeriodoConcessivo;
    }

    public Date getDataFinalDoPeriodoConcessivo() {
        return this.dataFinalDoPeriodoConcessivo;
    }

    public void setDataFinalDoPeriodoConcessivo(Date dataFinalDoPeriodoConcessivo) {
        this.dataFinalDoPeriodoConcessivo = dataFinalDoPeriodoConcessivo;
    }

    public Boolean getAbono() {
        return this.abono;
    }

    public void setAbono(Boolean abono) {
        this.abono = abono;
    }

    public Integer getQuantidadeDiasAbono() {
        return this.quantidadeDiasAbono;
    }

    public void setQuantidadeDiasAbono(Integer quantidadeDiasAbono) {
        this.quantidadeDiasAbono = quantidadeDiasAbono;
    }

    public SituacaoDaFeriasEnum getSituacao() {
        return this.situacao;
    }

    public void setSituacao(SituacaoDaFeriasEnum situacao) {
        this.situacao = situacao;
    }

    public Date getDataInicialDoPeriodoDeGozo1() {
        return this.dataInicialDoPeriodoDeGozo1;
    }

    public void setDataInicialDoPeriodoDeGozo1(Date dataInicialDoPeriodoDeGozo1) {
        this.dataInicialDoPeriodoDeGozo1 = dataInicialDoPeriodoDeGozo1;
    }

    public Date getDataFinalDoPeriodoDeGozo1() {
        return this.dataFinalDoPeriodoDeGozo1;
    }

    public void setDataFinalDoPeriodoDeGozo1(Date dataFinalDoPeriodoDeGozo1) {
        this.dataFinalDoPeriodoDeGozo1 = dataFinalDoPeriodoDeGozo1;
    }

    public Date getDataInicialDoPeriodoDeGozo2() {
        return this.dataInicialDoPeriodoDeGozo2;
    }

    public void setDataInicialDoPeriodoDeGozo2(Date dataInicialDoPeriodoDeGozo2) {
        this.dataInicialDoPeriodoDeGozo2 = dataInicialDoPeriodoDeGozo2;
    }

    public Date getDataFinalDoPeriodoDeGozo2() {
        return this.dataFinalDoPeriodoDeGozo2;
    }

    public void setDataFinalDoPeriodoDeGozo2(Date dataFinalDoPeriodoDeGozo2) {
        this.dataFinalDoPeriodoDeGozo2 = dataFinalDoPeriodoDeGozo2;
    }

    public Date getDataInicialDoPeriodoDeGozo3() {
        return this.dataInicialDoPeriodoDeGozo3;
    }

    public void setDataInicialDoPeriodoDeGozo3(Date dataInicialDoPeriodoDeGozo3) {
        this.dataInicialDoPeriodoDeGozo3 = dataInicialDoPeriodoDeGozo3;
    }

    public Date getDataFinalDoPeriodoDeGozo3() {
        return this.dataFinalDoPeriodoDeGozo3;
    }

    public void setDataFinalDoPeriodoDeGozo3(Date dataFinalDoPeriodoDeGozo3) {
        this.dataFinalDoPeriodoDeGozo3 = dataFinalDoPeriodoDeGozo3;
    }

    public Integer getPrazo() {
        return this.prazo;
    }

    public void setPrazo(Integer prazo) {
        this.prazo = prazo;
    }

    public Boolean getDobraGeral() {
        return this.dobraGeral;
    }

    public void setDobraGeral(Boolean dobraGeral) {
        this.dobraGeral = dobraGeral;
    }

    public Boolean getDobraDoPeriodoDeGozo1() {
        return this.dobraDoPeriodoDeGozo1;
    }

    public void setDobraDoPeriodoDeGozo1(Boolean dobraDoPeriodoDeGozo1) {
        this.dobraDoPeriodoDeGozo1 = dobraDoPeriodoDeGozo1;
    }

    public Boolean getDobraDoPeriodoDeGozo2() {
        return this.dobraDoPeriodoDeGozo2;
    }

    public void setDobraDoPeriodoDeGozo2(Boolean dobraDoPeriodoDeGozo2) {
        this.dobraDoPeriodoDeGozo2 = dobraDoPeriodoDeGozo2;
    }

    public Boolean getDobraDoPeriodoDeGozo3() {
        return this.dobraDoPeriodoDeGozo3;
    }

    public void setDobraDoPeriodoDeGozo3(Boolean dobraDoPeriodoDeGozo3) {
        this.dobraDoPeriodoDeGozo3 = dobraDoPeriodoDeGozo3;
    }

    @Override
    protected EntidadeBase validar() {
        GerenciadorDeValidadores.getInstance().validar(Ferias.class, this);
        return this;
    }

    @Override
    public void salvar() {
        super.salvar();
    }

    public void limparPeriodosDeGozos() {
        this.setDataInicialDoPeriodoDeGozo1(null);
        this.setDataFinalDoPeriodoDeGozo1(null);
        this.setDataInicialDoPeriodoDeGozo2(null);
        this.setDataFinalDoPeriodoDeGozo2(null);
        this.setDataInicialDoPeriodoDeGozo3(null);
        this.setDataFinalDoPeriodoDeGozo3(null);
    }

    public void sugerirPrazo() {
        int faltasNaoJustificadas = this.getCalculo().obterFaltasNaoJustificadas(this.getPeriodoAquisitivo());
        CalculoDoPrazoDeFerias calculoDoPrazoDeFerias = CalculoDoPrazoDeFerias.getInstance();
        calculoDoPrazoDeFerias.parametros(this.getPeriodoAquisitivo().getFinal(), this.calculo.getRegimeDoContrato(), faltasNaoJustificadas).executar();
        int prazo = calculoDoPrazoDeFerias.getResultado();
        this.setPrazo(prazo);
    }

    public void sugerirPrimeiroPeriodosDeGozos() {
        if (Utils.naoNulo(this.getCalculo().getInicioFeriasColetivas()) && HelperDate.dateBefore(this.getDataFinalDoPeriodoAquisitivo(), this.getCalculo().getInicioFeriasColetivas())) {
            Date dataInicial = this.getCalculo().getInicioFeriasColetivas();
            Periodo periodoDeGozo = new Periodo(dataInicial, HelperDate.getInstance(dataInicial).addDay(this.prazo - 1).getDate());
            this.setPeriodoDeGozo1(periodoDeGozo);
        } else if (SituacaoDaFeriasEnum.GOZADAS.equals((Object)this.getSituacao()) && Utils.naoNulos(this.dataFinalDoPeriodoConcessivo, this.prazo) && this.prazo > 0) {
            Date dataInicial = HelperDate.getInstance(this.dataFinalDoPeriodoConcessivo).addDay(-(this.prazo - 1)).getDate();
            Periodo periodoDeGozo = new Periodo(dataInicial, this.dataFinalDoPeriodoConcessivo);
            this.setPeriodoDeGozo1(periodoDeGozo);
        }
    }

    public String toString() {
        return super.toString("id", "relativa");
    }

    @Override
    public int hashCode() {
        return this.getHashCodeBuilder().append((Object)this.prazo).append((Object)this.situacao).append((Object)this.dobraGeral).append((Object)this.abono).append((Object)this.dataInicialDoPeriodoDeGozo1).append((Object)this.dataFinalDoPeriodoDeGozo1).append((Object)this.dobraDoPeriodoDeGozo1).append((Object)this.dataInicialDoPeriodoDeGozo2).append((Object)this.dataFinalDoPeriodoDeGozo2).append((Object)this.dobraDoPeriodoDeGozo2).append((Object)this.dataInicialDoPeriodoDeGozo3).append((Object)this.dataFinalDoPeriodoDeGozo3).append((Object)this.dobraDoPeriodoDeGozo3).hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (this.getClass() != obj.getClass()) {
            return false;
        }
        return this.getEqualsBuilder(obj).append((Object)this.prazo, (Object)((Ferias)obj).prazo).append((Object)this.situacao, (Object)((Ferias)obj).situacao).append((Object)this.dobraGeral, (Object)((Ferias)obj).dobraGeral).append((Object)this.abono, (Object)((Ferias)obj).abono).append((Object)this.dataInicialDoPeriodoDeGozo1, (Object)((Ferias)obj).dataInicialDoPeriodoDeGozo1).append((Object)this.dataFinalDoPeriodoDeGozo1, (Object)((Ferias)obj).dataFinalDoPeriodoDeGozo1).append((Object)this.dobraDoPeriodoDeGozo1, (Object)((Ferias)obj).dobraDoPeriodoDeGozo1).append((Object)this.dataInicialDoPeriodoDeGozo2, (Object)((Ferias)obj).dataInicialDoPeriodoDeGozo2).append((Object)this.dataFinalDoPeriodoDeGozo2, (Object)((Ferias)obj).dataFinalDoPeriodoDeGozo2).append((Object)this.dobraDoPeriodoDeGozo2, (Object)((Ferias)obj).dobraDoPeriodoDeGozo2).append((Object)this.dataInicialDoPeriodoDeGozo3, (Object)((Ferias)obj).dataInicialDoPeriodoDeGozo3).append((Object)this.dataFinalDoPeriodoDeGozo3, (Object)((Ferias)obj).dataFinalDoPeriodoDeGozo3).append((Object)this.dobraDoPeriodoDeGozo3, (Object)((Ferias)obj).dobraDoPeriodoDeGozo3).isEquals();
    }

    public static int encontrarPrazoFeriasProporcionais(ParametroDoTermo parametro) {
        if (Utils.naoNulo(parametro.getCalculo().getPrazoFeriasProporcional())) {
            return parametro.getCalculo().getPrazoFeriasProporcional();
        }
        int faltasNaoJustificadas = parametro.getCalculo().obterFaltasNaoJustificadas(parametro.getPeriodoAquisitivo());
        CalculoDoPrazoDeFerias calculoDoPrazoDeFerias = CalculoDoPrazoDeFerias.getInstance();
        calculoDoPrazoDeFerias.parametros(parametro.getPeriodoAquisitivo().getFinal(), parametro.getCalculo().getRegimeDoContrato(), faltasNaoJustificadas).executar();
        return calculoDoPrazoDeFerias.getResultado();
    }
}

